File: LeetCode/DataStruct/Python/Leetcode953-verifyingAnAlienDictionary.py
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        mp = defaultdict(int)
        for i in range(len(order)):
            mp[order[i]] = i
        for i in range(1,len(words)):
            l,r = words[i - 1],words[i]
            x,y = 0,0
            while x < len(l) and y < len(r):
                if mp[l[x]] > mp[r[y]]: return False
                if mp[r[y]] > mp[l[x]]: break
                x,y = x + 1,y + 1
            if x < len(l) and y == len(r): return False
        return True

File: LeetCode/DataStruct/Rust/Leetcode953-verifyingAnAlienDictionary.rs
use std::collections::HashMap;
impl Solution {
  pub fn is_alien_sorted(words: Vec<String>, order: String) -> bool {
    let mut mp = HashMap::new();
    for i in 0..order.len() {
      mp.insert(&order[i..i + 1], i);
    }
    for i in 1..words.len() {
      let (l, r) = (&words[i - 1], &words[i]);
      let (mut x, mut y) = (0, 0);
      while x < l.len() && y < r.len() {
        if mp[&l[x..x + 1]] > mp[&r[y..y + 1]] {
          return false;
        }
        if mp[&l[x..x + 1]] < mp[&r[y..y + 1]] {
          break;
        }
        x += 1;
        y += 1;
      }
      if x < l.len() && y == r.len() {
        return false;
      }
    }
    true
  }
}
